### PR TITLE
Back 26: Github trending repos and developers

### PR DESCRIPTION
**Pull request by 2015builder**
Included the ScrapeRepos file that can scrape trending repos and developers and get the readme files of the repo. Also has Alex's method to interact with the openai api.
The github and openai keys are needed for this one to work

### CODE REVIEW

File: ScrapeRepos.cjs [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+const cheerio = require('cheerio');
+const { time } = require('console');
+const showdown = require('showdown');
+const readline = require('readline').createInterface({
+  input: process.stdin,
+  output: process.stdout
+});
+
+
+async function getUserInput() {
```
Comment (DogukanGun) [inline ScrapeRepos.cjs:11]: Instead of user input please take the decision by params

File: ScrapeRepos.cjs [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+  });
+}
+
+/** Fetches the repo owner and name for each repo on the GitHub trending page
+ * @param userInput 0: daily; 1: weekly; 2: monthly
+ * @returns an array that stores alternatingly the owner and the name of each repo
+ */
+async function fetchRepos(userInput) {
+  try {
+    const timeModes = ['?since=daily', '?since=weekly', '?since=monthly'];
```
Comment (DogukanGun) [inline ScrapeRepos.cjs:25]: Could you define the time modes as variables. All different time modes could be different value.

File: ScrapeRepos.cjs [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    const response = await axios.get('https://github.com/trending' + timeModes[userInput]);
```
Comment (DogukanGun) [inline ScrapeRepos.cjs:26]: Could you define the path as general variable.

File: ScrapeRepos.cjs [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+*/
+async function getReadme(owner, name) {
+  const readmePaths = [`https://raw.githubusercontent.com/${owner}/${name}/release/readme.md`,
+                         `https://raw.githubusercontent.com/${owner}/${name}/dev/README.rst`,
+                        `https://raw.githubusercontent.com/${owner}/${name}/main/README.md`,
+                        `https://raw.githubusercontent.com/${owner}/${name}/master/README.md`];
+ 
+  for(i = 0; i < readmePaths.length; i++){
+    try{
+      const response = await axios.get(readmePaths[i])
```
Comment (DogukanGun) [inline ScrapeRepos.cjs:65]: This operation could be a function

File: ScrapeRepos.cjs [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+      // Use the converter object to convert Markdown to HTML to String:
+      const html = converter.makeHtml(content);
+      const html2 = html.toString();
+      const text = html2.replace(/<[^>]*>/g, '');
+      const text2 = text.replace(/\n\s+/g, '');
+      const text3 = text2.replace(/\//g, "");
+      return text3;
+    }
+    catch(error){
```
Comment (DogukanGun) [inline ScrapeRepos.cjs:77]: Please log the error message for now. Later we should store the error in database or another service

File: ScrapeRepos.cjs [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+    }
+  }
+  console.log("Couldn't locate the read me file")
```
Comment (DogukanGun) [inline ScrapeRepos.cjs:81]: I think this log message is for debugging. Could you delete this log message.

File: ScrapeRepos.cjs [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+}
+
+/**  This function imports the ReadMe.md file for a repository (if it can be located)
+* @param owner - owner of the repo
+* @param name - name of the repo
+* @returns a string containing the text of the repo
+*           returns null if the repo can't be located
+*/
+async function getReadme(owner, name) {
+  const readmePaths = [`https://raw.githubusercontent.com/${owner}/${name}/release/readme.md`,
```
Comment (DogukanGun) [inline ScrapeRepos.cjs:58]: Could you add command here? Personally I didn't understand what the paths are used for

File: ScrapeRepos.cjs [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+}
+
+/** Gets the repo's information via GitHub's GraphQL API
+ * @param query GraphQL query for the repo (including owner and name)
+ * @param authToken personal authorization token
+ * @returns the json data for the requested repo as by the graphql query
+ */
+async function getRepoInfo(query, authToken){
+  try{
+    const response = await  axios.post('https://api.github.com/graphql', {
```
Comment (DogukanGun) [inline ScrapeRepos.cjs:92]: Could you define the path as global variable?

File: ScrapeRepos.cjs [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+}
+
+//TODO needs documentation, returns and revision!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! 
+/** Creates the ELI5 of the repository by passing the readme to chatgpt
+ * @param readME readMe file of the repository
+ * @returns SHOULD return the summary of the repo
+ */
+async function openAIRequestTurbo(readME) {
+  //const readme = await convertReadMe(); //here something wrong with the await
+  const OPENAI_API_KEY = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX";
```
Comment (DogukanGun) [inline ScrapeRepos.cjs:115]: Key should not be stored here. This way is not secure. My suggestion is leaving like that for now and then creating a issue in linear to find the most secure way.

File: ScrapeRepos.cjs [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+        },
+        {
+          role: "user",
+          content: question,
+        },
+      ],
+    }),
+  });
+  const data = await response.json();
+  //const content = data.choices[0].message.content;
```
Comment (DogukanGun) [inline ScrapeRepos.cjs:142]: Please delete not used commands.

**Review by DogukanGun — CHANGES_REQUESTED** [review]

File: ScrapeRepos.cjs (hunk 2 above)
Comment (2015builder) [reply ScrapeRepos.cjs:25]: In which scenario do you think would the time modes have a different value? I mean they don't change over time do they?

File: ScrapeRepos.cjs (hunk 4 above)
Comment (2015builder) [reply ScrapeRepos.cjs:65]: But do you think this would be helpful? I don't believe that this specific string transformation would be used by any other service or class

File: ScrapeRepos.cjs (hunk 9 above)
Comment (2015builder) [reply ScrapeRepos.cjs:115]: Yes I totally agree with you.But for now (and testing purposes) we should leave it like that

File: ScrapeRepos.cjs (hunk 2 above)
Comment (DogukanGun) [reply ScrapeRepos.cjs:18]: When code becomes bigger it would hard to track which index represents which timeMode and those modes can be used by other modules therefore, they might be stored in common class.

File: ScrapeRepos.cjs (hunk 4 above)
Comment (DogukanGun) [reply ScrapeRepos.cjs:62]: This could help another developer who will develop something in this class. Sometimes functions are used to make code more readable. My purpose at this review was about readability.

File: ScrapeRepos.cjs (hunk 9 above)
Comment (DogukanGun) [reply ScrapeRepos.cjs:110]: Yes I agree with you, please don't forget to open a new issue for this.

File: ScrapeRepos.cjs (hunk 2 above)
Comment (2015builder) [reply ScrapeRepos.cjs:18]: I will implement this in my next PR

File: ScrapeRepos.cjs (hunk 4 above)
Comment (2015builder) [reply ScrapeRepos.cjs:62]: I will also implement this in next PR

File: ScrapeRepos.cjs (hunk 9 above)
Comment (2015builder) [reply ScrapeRepos.cjs:110]: I opened the issue on Linear and also added a discussion item to the sprint meeting
